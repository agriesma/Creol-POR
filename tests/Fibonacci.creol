class Fibonacci(n:Integer)
begin
  op init == skip

  op run == var res: Integer; t!fib(n); await t?; t?(res)

  op fib(in n:Integer; out f:Integer) ==
    var a:Integer; var b:Integer;
    if n = 0 then f := 1 else
      if n = 1 then f := 1 else
  	s!fib(n - 1); s?(a); t!fib(n - 2); t?(b); f := a + b
      fi
    fi
with Any
  op fibonacci(in n:Integer; out f:Integer) == t!fib(n); t?(f)
end

class FibTester(f: Fibonacci, n: Integer)
begin
	op init == skip
	op run ==
	 	var res:Integer;
 		t!f.fibonacci(n);
 		t?(res);
end
