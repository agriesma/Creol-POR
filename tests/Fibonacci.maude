load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Fibonacci" : Cl | Inh: noInh, Par: "n", Att: "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: "res" |-> null, Code: "t" ! "this" . "fib" ( "n" ) ; ( await ( "t" ?? ) ) ; ( "t" ? ( "res" ) )  ; return ( emp ) > *
  < "fib" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null , "a" |-> null , "b" |-> null, Code: if "=" ( "n" # int(0) ) th "f" ::= int(1) el if "=" ( "n" # int(1) ) th "f" ::= int(1) el "s" ! "this" . "fib" ( "-" ( "n" # int(1) ) ) ; ( "s" ? ( "a" ) )  ; "t" ! "this" . "fib" ( "-" ( "n" # int(2) ) ) ; ( "t" ? ( "b" ) )  ; "f" ::= "+" ( "a" # "b" ) fi fi ; return ( "f" ) > *
  < "fibonacci" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null, Code: "t" ! "this" . "fib" ( "n" ) ; ( "t" ? ( "f" ) )  ; return ( "f" ) >, Ocnt: 0 >

< "FibTester" : Cl | Inh: noInh, Par: "f" , "n", Att: "f" |-> null , "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: "res" |-> null, Code: "f" . "fibonacci" ( "n" ; "res" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
