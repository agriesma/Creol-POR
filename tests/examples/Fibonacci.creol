interface Fibonacci
begin
  with Any
    op fibonacci(in n: Int; out f: Int)
end

class Fibonacci(n:Int)
begin
  op init == skip

  op run == var res: Int; var t: Label[Int]; t!fib(n); await t?; t?(res)

  op fib(in n:Int; out f:Int) ==
    var a:Int; var b:Int; var s: Label[Int]; var t:Label[Int];
    if n = 0 then f := 1 else
      if n = 1 then f := 1 else
  	s!fib(n - 1); t!fib(n - 2); s?(a); t?(b); f := a + b
      end
    end

with Any
  op fibonacci(in n:Int; out f:Int) == var t:Label[Int]; t!fib(n); t?(f)
end

class FibTester(f: Fibonacci, n: Int)
begin
	op init == skip
	op run ==
	 	var res:Int;
 		f.fibonacci(n; res)
end
