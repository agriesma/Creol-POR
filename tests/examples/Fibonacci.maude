load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Fibonacci" : Cl | Inh: noInh, Par: "n", Att: "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "res" |-> null , "t" |-> null, Code: ( "t" ! "this" . "fib" ("n" ) )  ; ( await ( "t" ?? ) ) ; ( "t" ? ( "res" ) )  ; return ( emp ) > *
  < "fib" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null , "a" |-> null , "b" |-> null , "s" |-> null , "t" |-> null, Code: if "=" ( "n" # int(0) ) th "f" ::= int(1) el if "=" ( "n" # int(1) ) th "f" ::= int(1) el ( "s" ! "this" . "fib" ("-" ( "n" # int(1) ) ) )  ; ( "t" ! "this" . "fib" ("-" ( "n" # int(2) ) ) )  ; ( "s" ? ( "a" ) )  ; ( "t" ? ( "b" ) )  ; "f" ::= "+" ( "a" # "b" ) fi fi ; return ( "f" ) > *
  < "fibonacci" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null , "t" |-> null, Code: ( "t" ! "this" . "fib" ("n" ) )  ; ( "t" ? ( "f" ) )  ; return ( "f" ) >, Ocnt: 0 >

< "FibTester" : Cl | Inh: noInh, Par: "f" , "n", Att: "f" |-> null , "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "res" |-> null, Code: "sync.0" ! "f" . "fibonacci" ( "n" ) ; ( "sync.0" ? ( "res" ) )  ; return ( emp ) >, Ocnt: 0 >

.
endm
