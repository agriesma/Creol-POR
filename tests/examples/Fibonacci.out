==========================================
reduce in PROGRAM : classes .
result Configuration: 
< "FibTester" : Class | Inh: noInh, Param: "n", Att: "f" |-> null, "n" |->
    null, "res" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: new("f" ;
    "Fibonacci" ; "n") ; return(emp) >, < "run" : Method | Param: noVid, Att:
    "_" |-> null, "label:0" |-> null, Code: call("label:0" ; "f" ; "fibonacci"
    ; "n") ; get("label:0" ; "res") ; return(emp) >,
  Ocnt: 0 >
 
< "Fibonacci" : Class | Inh: noInh, Param: "n", Att: "n" |-> null, Mtds:
  < "fib" : Method | Param: "n", Att: "_" |-> null, "a" |-> null, "b" |-> null,
    "f" |-> null, "n" |-> null, "s" |-> null, "t" |-> null, Code: if "="("n" ::
    int(0)) th assign("f" ; int(1)) el if "="("n" :: int(1)) th assign("f" ;
    int(1)) el call("s" ; "this" ; "fib" ; "-"("n" :: int(1))) ; call("t" ;
    "this" ; "fib" ; "-"("n" :: int(2))) ; get("s" ; "a") ; get("t" ; "b") ;
    assign("f" ; "+"("a" :: "b")) fi fi ; return("f") >, < "fibonacci" : Method
    | Param: "n", Att: "_" |-> null, "f" |-> null, "n" |-> null, "t" |-> null,
    Code: call("t" ; "this" ; "fib" ; "n") ; get("t" ; "f") ; return("f") >, <
    "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "res" |-> null, "t"
    |-> null, Code: call("t" ; "this" ; "fib" ; "n") ; await ?("t") ; get("t" ;
    "res") ; return(emp) >,
  Ocnt: 0 >

Bye.
